File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string str) {
        if(str.length() % 2 != 0) return false;
        std::stack<char> s;

        for(const auto& c : str){
            if(c == '(' || c == '{' || c == '[')
                s.push(c);
            else if(s.empty() ||
                    (c == ')' && s.top() != '(') || 
                    (c == '}' && s.top() != '{') || 
                    (c == ']' && s.top() != '[' )
                    )
                return false;
            else s.pop();
        }

        return s.empty();
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.go
func isValid(s string) bool {
    stack := []rune{}

    for _, char := range s{
        if char == '(' || char == '{' || char == '['{
            stack = append(stack, char)
        }else{
            if len(stack) == 0{
                return false
            }
            last := stack[len(stack)-1]
            stack = stack[:len(stack)-1]
            c := string(last) + string(char)
            if c == "()" || c == "{}" || c == "[]"{
                continue
            }else{
                return false
            }
        }
    }

    return len(stack) == 0
}